kie-server-parent/kie-server-controller/kie-server-controller-impl/src/main/java/org/kie/server/controller/impl/KieServerControllerAdminImpl.java

```diff
@@ -60,55 +74,91 @@ public KieServerInstance addKieServerInstance(KieServerInfo kieServerInfo) throw
             KieServerInstanceInfo instanceInfo = new KieServerInstanceInfo(kieServerInfo.getLocation(), KieServerStatus.UP, kieServerInfo.getCapabilities());
             kieServerInstance.getManagedInstances().add(instanceInfo);
         }
-        storage.store(kieServerInstance);
+        // persist the configuration in based on server templates
+        serverTemplate = new ServerTemplate();
+        serverTemplate.setId(kieServerInfo.getServerId());
+        serverTemplate.setName(kieServerInfo.getName());
+        serverTemplate.setCapabilities(kieServerInfo.getCapabilities());
+
+        if (kieServerInfo.getLocation() != null && !kieServerInfo.getLocation().isEmpty()) {
+            serverTemplate.addServerInstance(ModelFactory.newServerInstanceKey(serverTemplate.getId(), kieServerInfo.getLocation()));
+        }
+
+        templateStorage.store(serverTemplate);
+
+        // return kie server instance for backward compatibility reason
         return kieServerInstance;
     }
 
     @Override
     public KieServerInstance removeKieServerInstance(String identifier) throws KieServerControllerException {
-        KieServerInstance kieServerInstance = this.storage.load(identifier);
-        if (kieServerInstance == null) {
+
+        if (!this.templateStorage.exists(identifier)) {
             throw new KieServerControllerException("KieServerInstance not found with id: " + identifier);
         }
-        return storage.delete(identifier);
+        templateStorage.delete(identifier);
+        return null;
     }
 
     @Override
     public List<KieServerInstance> listKieServerInstances() {
-        return storage.load();
+        List<KieServerInstance> serverInstances = new ArrayList<KieServerInstance>();
+        Collection<ServerTemplate> templates = templateStorage.load();
+
+        for (ServerTemplate serverTemplate : templates) {
+            KieServerInstance kieServerInstance = forServerTemplate(serverTemplate);
+
+            serverInstances.add(kieServerInstance);
+        }
+
+
+        return serverInstances;
     }
 
     @Override
     public KieServerInstance getKieServerInstance(String identifier) throws KieServerControllerException {
-        KieServerInstance kieServerInstance = this.storage.load(identifier);
-        if (kieServerInstance == null) {
+        ServerTemplate serverTemplate = this.templateStorage.load(identifier);
+        if (serverTemplate == null) {
```

Comment:
seems like load(id) could return null that breaks duplicated server to no existent id.